lib/policy_machine_storage_adapters/active_record.rb

```diff
@@ -290,6 +317,38 @@ def assert_valid_attributes!(attributes)
           raise ArgumentError, "Provided argument contains invalid keys, valid keys are #{PolicyElement.column_names}"
         end
       end
+
+      # Returns a hash containing the ancestors of the root nodes, with the interstitial
+      # ancestor relationships represented by the key/value pairs.
+      def get_ancestor_id_tree(root_nodes)
+        Assignment.select_ancestor_ids(root_nodes).each_with_object({}) do |row, memo|
+          id_array = row['ancestor_ids'].tr('{}','').split(',')
+          id_array.each { |ancestor_id| memo[ancestor_id] ||= [] }
+          memo[row['id']] = id_array
+        end
+      end
+
+      # Zip an array of attributes into an id tree, with interstitial relationships preserved
+      # under the "relative_ids" key
+      def zip_attributes_into_id_tree!(id_tree, plucked_fields, plucked_attributes)
```

Comment:
Why mutate `id_tree` in place? Can we not do that? I'd prefer to just set the return value to the changed id tree?